File: providers/aws/src/main/java/cloudspec/aws/ec2/resource/nested/EC2InternetGatewayAttachment.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.resource.nested;

import cloudspec.annotation.AssociationDefinition;
import cloudspec.annotation.PropertyDefinition;
import cloudspec.aws.ec2.resource.EC2VpcResource;
import software.amazon.awssdk.services.ec2.model.InternetGatewayAttachment;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class EC2InternetGatewayAttachment {
    @PropertyDefinition(
            name = "state",
            description = "The current state of the attachment",
            exampleValues = "attaching | attached | detaching | detached"
    )
    private final String state;

    @AssociationDefinition(
            name = "vpc",
            description = "The VPC",
            targetClass = EC2VpcResource.class
    )
    private final String vpcId;

    public EC2InternetGatewayAttachment(String state, String vpcId) {
        this.state = state;
        this.vpcId = vpcId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (o == null || (getClass() != o.getClass() && !(o instanceof InternetGatewayAttachment))) {
            return false;
        }

        if (o instanceof InternetGatewayAttachment) {
            return sdkEquals((InternetGatewayAttachment) o);
        }

        EC2InternetGatewayAttachment that = (EC2InternetGatewayAttachment) o;
        return Objects.equals(state, that.state) &&
                Objects.equals(vpcId, that.vpcId);
    }

    private boolean sdkEquals(InternetGatewayAttachment that) {
        return Objects.equals(state, that.stateAsString()) &&
                Objects.equals(vpcId, that.vpcId());
    }

    @Override
    public int hashCode() {
        return Objects.hash(state, vpcId);
    }

    public static List<EC2InternetGatewayAttachment> fromSdk(List<InternetGatewayAttachment> internetGatewayAttachments) {
        return Optional.ofNullable(internetGatewayAttachments)
                       .orElse(Collections.emptyList())
                       .stream()
                       .map(EC2InternetGatewayAttachment::fromSdk)
                       .collect(Collectors.toList());
    }

    public static EC2InternetGatewayAttachment fromSdk(InternetGatewayAttachment internetGatewayAttachment) {
        return Optional.ofNullable(internetGatewayAttachment)
                       .map(v ->
                               new EC2InternetGatewayAttachment(
                                       v.stateAsString(),
                                       v.vpcId()
                               )
                       )
                       .orElse(null);
    }
}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/nested/EC2InternetGatewayAttachment.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2.nested

import cloudspec.annotation.AssociationDefinition
import cloudspec.annotation.PropertyDefinition
import cloudspec.aws.ec2.EC2Vpc

data class EC2InternetGatewayAttachment(
        @property:PropertyDefinition(
                name = PROP_STATE,
                description = PROP_STATE_D,
                exampleValues = "attaching | attached | detaching | detached"
        )
        val state: String?,

        @property:AssociationDefinition(
                name = ASSOC_VPC,
                description = ASSOC_VPC_D,
                targetClass = EC2Vpc::class
        )
        val vpcId: String?
) {
    companion object {
        const val PROP_STATE = "state"
        const val PROP_STATE_D = "The current state of the attachment"
        const val ASSOC_VPC = "vpc"
        const val ASSOC_VPC_D = "The VPC"
    }
}
